### PR TITLE
dtypes and arrays scalars

### PR DESCRIPTION
**Pull request by ev-br**
Address gh-9: implement the scalar type and array dtype hierarchy.

Implementation-wise it does a fairly weird thing of overriding `__new__` so that e.g.
`np.float32(1)` constructs a zero-dim tensor, *not*  a `np.float32` instance. 

See also https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/5#issuecomment-1371860957 and below for a discussion about array scalars.

closes gh-9 when done.

### CODE REVIEW

**Comment by lezcano** [comment]
nit. Why don't you submit this as a merge into `reductions` to get a cleaner diff in the whole PR?

**Comment by ev-br** [comment]
Note that the initial review is at gh-13.

**Comment by ev-br** [comment]
OK, this works up to `test_scalarmath`. 
Fixing that one involves implementing a decent subset of ufuncs, so I'm going to take it to a separate branch.

File: .gitignore [hunk 1]
```diff
@@ -3,5 +3,6 @@ autogen/__pycache__
 torch_np/__pycache__/*
 torch_np/tests/__pycache__/*
 torch_np/tests/numpy_tests/core/__pycache__/*
+torch_np/testing/__pycache__/*
```
Comment (lezcano) [inline .gitignore:6]:
```suggestion
**/__pycache__/
```
Let's just ignore them all.

Comment (lezcano) [reply .gitignore:6]: Even more, a better syntax may be
```python
# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
```
as per the [oficial gitignore](https://github.com/github/gitignore/blob/main/Python.gitignore).

**Comment by ev-br** [comment]
Ok, thanks for the reviews and discussions, let's merge this and iterate.

**Pull request merged**